docs/examples/intro_to_kernels.py

```diff
@@ -0,0 +1,328 @@
+#
+# $$ \begin{aligned}
+# \log(p(\mathbf{y} | \mathbf{X}, \boldsymbol{\theta})) &= \log\left(\int p(\mathbf{y} | \mathbf{f}, \mathbf{X}, \boldsymbol{\theta}) p(\mathbf{f} | \mathbf{X}, \boldsymbol{\theta}) d\mathbf{f}\right) \nonumber \\
+# &= - \frac{1}{2} \mathbf{y} ^ \top \left(K(\mathbf{X}, \mathbf{X}) + \sigma^2 \mathbf{I} \right)^{-1} \mathbf{y} - \frac{1}{2} \log |K(\mathbf{X}, \mathbf{X}) + \sigma^2 \mathbf{I}| - \frac{n}{2} \log 2 \pi
+# \end{aligned}$$
+
+# %% [markdown]
+# We'll demonstrate the advantages of being able to infer kernel parameters from the training data by fitting a GP to the widely used [Forrester function](https://www.sfu.ca/~ssurjano/forretal08.html):
+#
+# $$f(x) = (6x - 2)^2 \sin(12x - 4)$$
+
+
+# %%
+# Forrester function
+def forrester(x):
+    return (6 * x - 2) ** 2 * jnp.sin(12 * x - 4)
+
+
+n = 5
+
+training_x = jr.uniform(key=key, minval=0, maxval=1, shape=(n,)).reshape(-1, 1)
+training_y = forrester(training_x)
+D = gpx.Dataset(X=training_x, y=training_y)
+
+test_x = jnp.linspace(0, 1, 100).reshape(-1, 1)
+test_y = forrester(test_x)
+
+# %% [markdown]
+# First we define our model, using the Matérn32 kernel, and construct our posterior *without* optimising the kernel hyperparameters:
+
+# %%
+mean = gpx.mean_functions.Zero()
+kernel = gpx.kernels.Matern32(
+    lengthscale=jnp.array(2.0)
+)  # Initialise our kernel lengthscale to 2.0
+
+prior = gpx.Prior(mean_function=mean, kernel=kernel)
+
+likelihood = gpx.Gaussian(
+    num_datapoints=D.n, obs_noise=jnp.array(1e-6)
+)  # Our function is noise-free, so we set the observation noise to a very small value
+likelihood = likelihood.replace_trainable(obs_noise=False)
+
+no_opt_posterior = prior * likelihood
+
+# %% [markdown]
+# We can then optimise the hyperparmeters by minimising the negative log marginal likelihood of the data:
+
+# %%
+negative_mll = gpx.objectives.ConjugateMLL(negative=True)
+negative_mll(no_opt_posterior, train_data=D)
+negative_mll = jit(negative_mll)
+
+opt_posterior, history = gpx.fit(
+    model=no_opt_posterior,
+    objective=negative_mll,
+    train_data=D,
+    optim=ox.adam(learning_rate=0.01),
+    num_iters=2000,
+    safe=True,
+    key=key,
+)
+
+
```

Comment:
A short markdown cell here stating that you'll now query the predictive posterior.

Reply:
No need for maths, just break up a code block with some narration.